File: client/src/main/java/com/noqapp/android/client/views/adapters/AddressListAdapter.java
package com.noqapp.android.client.views.adapters;


import android.app.Activity;
import android.graphics.Color;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.noqapp.android.client.R;
import com.noqapp.android.client.utils.ShowCustomDialog;
import com.noqapp.android.common.beans.JsonUserAddress;

import java.util.List;

public class AddressListAdapter extends ArrayAdapter<JsonUserAddress> {
    private List<JsonUserAddress> list;
    private LayoutInflater inflator;
    private Activity context;
    private UpdateAddress updateAddress;
    private String userAddressId;

    public AddressListAdapter(Activity context, List<JsonUserAddress> list, UpdateAddress updateAddress, String userAddressId) {
        super(context, R.layout.list_item_selected, list);
        this.list = list;
        inflator = context.getLayoutInflater();
        this.context = context;
        this.updateAddress = updateAddress;
        this.userAddressId = userAddressId;
    }

    @Override
    public View getView(final int position, View convertView, ViewGroup parent) {
        ViewHolder holder;
        if (convertView == null) {
            convertView = inflator.inflate(R.layout.list_item_selected, null);
            holder = new ViewHolder();
            holder.title = convertView.findViewById(R.id.title);
            holder.tv_primary = convertView.findViewById(R.id.tv_primary);
            holder.iv_delete = convertView.findViewById(R.id.iv_delete);
            convertView.setTag(holder);
            convertView.setTag(R.id.title, holder.title);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

        holder.title.setText(list.get(position).getAddress());
        if (list.get(position).getId().equals(userAddressId)) {
            holder.tv_primary.setText("Primary");
            holder.tv_primary.setTextColor(Color.BLACK);
        } else {
            holder.tv_primary.setText("Set Primary");
            holder.tv_primary.setTextColor(Color.LTGRAY);
        }
        holder.tv_primary.setOnClickListener(v -> {
            if (list.get(position).getId().equals(userAddressId)) {
                // do nothing
            } else {
                if (null != updateAddress)
                    updateAddress.setPrimaryAddress(list.get(position));
            }
        });
        holder.iv_delete.setOnClickListener((View v) -> {
            ShowCustomDialog showDialog = new ShowCustomDialog(context, true);
            showDialog.setDialogClickListener(new ShowCustomDialog.DialogClickListener() {
                @Override
                public void btnPositiveClick() {
                    if (null != updateAddress)
                        updateAddress.removeAddress(list.get(position));
                }

                @Override
                public void btnNegativeClick() {
                    //Do nothing
                }
            });
            showDialog.displayDialog("Delete Address", "Do you want to delete address from address list?");
        });

        return convertView;
    }

    static class ViewHolder {
        private TextView title;
        private TextView tv_primary;
        private ImageView iv_delete;
    }

    public interface UpdateAddress {
        void removeAddress(JsonUserAddress jsonUserAddress);

        void setPrimaryAddress(JsonUserAddress jsonUserAddress);
    }

}

File: client/src/main/java/com/noqapp/android/client/views/adapters/AddressListAdapter.kt
package com.noqapp.android.client.views.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RadioButton
import android.widget.TextView
import androidx.appcompat.widget.AppCompatImageView
import androidx.recyclerview.widget.RecyclerView
import com.noqapp.android.client.R
import com.noqapp.android.client.databinding.AddressListItemBinding
import com.noqapp.android.client.utils.ShowCustomDialog
import com.noqapp.android.client.views.activities.NoqApplication
import com.noqapp.android.common.beans.JsonUserAddress

/**
 * Adapter class for showing address list
 * Created by Vivek Jha on 21/03/2021
 */
class AddressListAdapter(private val addressList: MutableList<JsonUserAddress>, val clickListener: (JsonUserAddress, View) -> Unit) : RecyclerView.Adapter<AddressListAdapter.AddressListViewHolder>() {

    inner class AddressListViewHolder(private val addressListItemView: View) : RecyclerView.ViewHolder(addressListItemView) {
        private val tvPrimary: TextView
        private val tvAddress: TextView
        private val ivDelete: AppCompatImageView
        private val viewDivider: View
        private val rbAddress: RadioButton

        init {
            val binding = AddressListItemBinding.bind(addressListItemView)
            tvPrimary = binding.tvPrimary
            tvAddress = binding.tvAddress
            ivDelete = binding.ivDelete
            viewDivider = binding.viewDivider
            rbAddress = binding.rbAddress
        }

        fun bind(jsonUserAddress: JsonUserAddress, position: Int) {
            tvAddress.text = jsonUserAddress.address

            if (jsonUserAddress.isPrimaryAddress) {
                tvPrimary.visibility = View.VISIBLE
            } else {
                tvPrimary.visibility = View.GONE
            }

            if (position == addressList.size - 1) {
                viewDivider.visibility = View.GONE
            } else {
                viewDivider.visibility = View.VISIBLE
            }

            ivDelete.setOnClickListener {
                val showDialog = ShowCustomDialog(it.context, true)
                showDialog.setDialogClickListener(object : ShowCustomDialog.DialogClickListener {
                    override fun btnPositiveClick() {
                        clickListener(jsonUserAddress, it)
                    }

                    override fun btnNegativeClick() {
                        //Do nothing
                    }
                })
                showDialog.displayDialog("Delete Address", "Do you want to delete address from address list?")
            }

            rbAddress.isChecked = jsonUserAddress.id == NoqApplication.getSelectedAddressId()

            rbAddress.setOnClickListener {
                clickListener(jsonUserAddress, it)
            }

            addressListItemView.setOnClickListener {
                clickListener(jsonUserAddress, it)
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AddressListViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.address_list_item, parent, false)
        return AddressListViewHolder(view)
    }

    override fun getItemCount(): Int {
        return addressList.size
    }

    override fun onBindViewHolder(holder: AddressListViewHolder, position: Int) {
        holder.bind(addressList[position], position)
    }

    fun addItems(jsonUserAddressList: List<JsonUserAddress>) {
        addressList.clear()
        addressList.addAll(jsonUserAddressList)
        notifyDataSetChanged()
    }
}
